### DISCUSSION TITLE
Pixel Choke.

### DISCUSSION THREAD

**Discussion by anthonypaulbalducci** [ImageMagick/ImageMagick#5571] [Help] [answered]
I should start of by saying this is *such* a great/cool library.

But there is something I've wanted/needed to do but haven't quite figured it out yet.

I've seen Fred's excellent scripts, particularly his 'Magic Wand' implementation, which could 'get me close'--

I could, insist the user already provide a transparent background in the image to be processed, which would be 'much easier'.

But basically, what I need, is just as in Photoshop, 'select solid background' --> Invert (Delete) to give you your image selection --> And then the 'contract selection' option, by 'x' number of pixels.

From there, I am good, but I haven't figured out how to do it yet.

Any advice, appreciated.

Best,
-A

**Comment by fmw42** [comment]
Can you provide an example image (input and PS result)?

As a guess, I suspect you need (Imagemagick 6 syntax)

`convert image -fuzz XX% -transparent bgcolor -channel a <-negate> -morphology erode diamond:X +channel result`

where you provide the bgcolor (background color) and X is approximately the number of pixels to contract (erode). The XX% you provide would be some small number in percent that you want to consider near the exact background color to include making transparent. If you have a perfectly flat (constant) background, the XX could be 0.

I am not sure in Imagemagick you want the invert, so here is an example without it. (If I use the negate, then you would just get the "sky".)

![boat](https://user-images.githubusercontent.com/13458218/190703051-fa2c9784-fb96-4cbc-a6cb-ebc59a02a60f.jpg)

`convert boat.jpg -fuzz 15% -transparent white -channel a -morphology erode diamond:3 +channel boat_result.png`

![boat_result](https://user-images.githubusercontent.com/13458218/190703139-8dac5d0d-0c67-4ba4-ac61-f489250a2062.png)

If on Imagemagick 7, use magick in place of convert

**Reply by anthonypaulbalducci** [reply]
Oh hey,

So 'Fred himself' responded, and to be honest, I got a little nervous...

But am wiling to be honest in my application. 

You may not be aware, but what is common in DTG or other forms of garment printing is, aside from the 'crop', you need to produce an alpha layer to lay down the 'white ink'.

Probably, only because the print precision is never 'perfect', so you accept some 'overbleed' in the final result.

And 'slaves', every day, workshops over perform this transform/result.

Yet, my thought is could not this be automated ?!?

![image](https://user-images.githubusercontent.com/34378880/190832984-206a6763-c6cc-4684-9fb5-ad0312a19a7c.png)

However, I am not sure. 

Putting a 'fuzz' is not 'quite the same' as 'contracting the mask'-- I just want to take an image, step exactly two steps back from it, and build my alpha channel.

Anyways Fred, sorry for my 'nervousness', but thanks so much for your help.

Best,
-A

**Reply by anthonypaulbalducci** [reply]
![image](https://user-images.githubusercontent.com/34378880/190833277-97283b83-d32b-4e6d-973e-fecdda3b53ed.png)

**Reply by anthonypaulbalducci** [reply]
And just so you know, this is the end result I have running: http://www.makki-usa.com/jeff/dtest/webgl_loader_obj_mtl.html
But it requires a lot of 'IM' tricks to make the texture right, cut, crop, get there....

**Reply by fmw42** [reply]
Sorry, but perhaps I do not understand what you mean by "contracting the mask" if not the same as the -morphology erode?  By the way, it is not the fuzz that is doing the "contracting" but the morphology erode. But it does not work by specifying an exact thickness. The erode amount is controlled by both the type of kernel (diamond, square, circle, etc) and the filter size. For a diamond, increasing the size by 1 is nearly the same as a thickness change of one.

**Reply by anthonypaulbalducci** [reply]
So I have sanitized it. Let's say this is my 'cutting mask' (I've already figured that part out).
![image](https://user-images.githubusercontent.com/34378880/190834198-cdd968a7-f950-4d2f-9d3c-e1e16f27a6a2.png)

But 'this' is the operation I need to perform. 

![image](https://user-images.githubusercontent.com/34378880/190834226-5671d1ee-ab1f-43ba-8854-e020d0fc4d10.png)

Irregardless of color.

I mean, now that I think about it... Maybe 'dumb me', you just make 'two masks', one smaller than the other. 

However, not all examples will be so simple, and cover 'the entire' garment.

I'm thinking of making this 'open source' so check out the rest of my work at allpointsbulletin.info.

But thanks so much for your help.

-A

**Comment by fmw42** [comment]
But you have not explained what contract does?  Please clarify. I do not see why it is not the same as -morphology erode or dilate depending upon your mask polarity.  Does it remove pixels from the outer boundary to make the mask "thinner"? If so, that is morphology erode if the mask is white on black and dilate if the mask is black on white. In IM, alpha channel masks are white on black (opposite of your PS mask), which is why I removed the -negate (invert) from the command.

**Comment by fmw42** [comment]
You can also make the mask thinner (contract?) by using distance transform (-morphology distance). See https://imagemagick.org/Usage/morphology/#distance.

**Reply by anthonypaulbalducci** [reply]
Dear Fred,

Well you are much more 'expert' at this than I am.

I figured the contraction operation, of a given selection, would be a simple 'subtraction' of the vector outlined. 

I have to be honest I have no idea how ImageMagick's point system works. But if it were centered at x,y on the center of the image, then you kind of just 'pull back' x number of pixels from the coordinate system.

If, though, the point system operates from the traditional 'top left', okay, I can see, this is 'way harder'.

I'll post some more images in the morning to make this clearer.

My 'covering the whole shirt' design makes it pretty simple, but there are harder examples.

And I'll read up on your morphology reference-- But I am a little bit like 'geez', I didn't realize it was so complicated !

Thanks again though.

Best,
-A

**Reply by anthonypaulbalducci** [reply]
Also, I've had another thought 'stirring in my brain'-- But the math is too hard for me. I think it 'should' be possible, but I've never seen it done, in implementation. The thought comes from my time working at Ricoh and relates to digital printing. I'll ask you later, once I have a sketch/drawing.

**Reply by fmw42** [reply]
Imagemagick is a raster processor. There is no vector outline. Even if there were, “pulling back” is still not easy in a vector system.

**Reply by anthonypaulbalducci** [reply]
Hey Fred, no, I follow you, or what you are saying. I guess I have just 'dived down the rabbit hole', proverbially speaking.

That said, this is kind of a 'for fun' or open source/small sales project.

My other idea I think though is better (commercial) and could use some 'help', if it works.

I had already tried yours, but drop me an e-mail at anthony.balducci@alumni.utoronto.ca, and I will explain.

Best,
-A

**Comment by snibgo** [comment]
@anthonypaulbalducci: You have shown a mask, and a "contracted" mask. But you have shown these as screen grabs of the program you have used. 

Please post the actual image files (probably PNG files). Then we can try to replicate the operation with IM.

**Comment by fmw42** [comment, marked as answer]
Here are two ways to "contract" a mask (5 px).

Mask:

![mask](https://user-images.githubusercontent.com/13458218/190868644-80ed55e6-6e81-4da6-988e-fd73b4fca934.png)

Morphology erode diamond:x

`magick mask.png -morphology erode diamond:5 mask_erode5.png`

![mask_erode5](https://user-images.githubusercontent.com/13458218/190868659-d84880ea-cfb3-4060-bd98-3b2137c11495.png)


Threshold morphology distance transform:

```
magick mask.png -threshold 50% +depth -morphology Distance Euclidean:4 x.png
magick identify -format 'Maximum Distance = %[max]' x.png

Maximum Distance = 12264 (distances scaled by 100)
5px = 500 units, so 500/12264 = 0.039107561458311

magick x.png -auto-level -threshold 3.9107561458311% tshirt_euclidean5.png
```

![tshirt_euclidean5](https://user-images.githubusercontent.com/13458218/190868765-7e59a946-ec97-4332-9ac4-cc0c2e5ac5d4.png)

These two results are close with slight differences in shape.  You be the judge.
